zcash_client_backend/src/scanning.rs

```diff
@@ -345,41 +340,106 @@ pub(crate) fn scan_block_with_runner<
+                                        at_height: cur_height,
+                                    })
+                                }
+                            },
+                        )
+                },
+                |m| {
+                    let sapling_output_count: u32 = block
+                        .vtx
+                        .iter()
+                        .map(|tx| tx.outputs.len())
+                        .sum::<usize>()
+                        .try_into()
+                        .expect("Sapling output count cannot exceed a u32");
+
+                    // The default for m.sapling_commitment_tree_size is zero, so we need to check
+                    // that the subtraction will not underflow; if it would do so, we given invalid
+                    // chain metadata for a block with Sapling outputs.
+                    if m.sapling_commitment_tree_size >= sapling_output_count {
+                        Ok(m.sapling_commitment_tree_size - sapling_output_count)
+                    } else {
+                        Err(ScanError::TreeSizeInvalid {
+                            protocol: ShieldedProtocol::Sapling,
+                            at_height: cur_height,
+                        })
+                    }
+                },
+            )
+        },
+        Ok,
+    )?;
+
+    let initial_orchard_tree_size = prior_block_metadata.and_then(|m| m.orchard_tree_size());
+    let mut orchard_commitment_tree_size = initial_orchard_tree_size.map_or_else(
+        || {
+            block.chain_metadata.as_ref().map_or_else(
+                || {
+                    // If we're below Orchard activation, or Orchard activation is not set, the tree size is zero
+                    params.activation_height(NetworkUpgrade::Nu5).map_or_else(
+                        || Ok(0),
+                        |orchard_activation| {
+                            if cur_height < orchard_activation {
+                                Ok(0)
+                            } else {
+                                Err(ScanError::TreeSizeUnknown {
+                                    protocol: ShieldedProtocol::Orchard,
+                                    at_height: cur_height,
+                                })
+                            }
+                        },
+                    )
+                },
+                |m| {
+                    let orchard_action_count: u32 = block
+                        .vtx
+                        .iter()
+                        .map(|tx| tx.actions.len())
+                        .sum::<usize>()
+                        .try_into()
+                        .expect("Orchard action count cannot exceed a u32");
+
+                    // The default for m.orchard_commitment_tree_size is zero, so we need to check
+                    // that the subtraction will not underflow; if it would do so, we given invalid
+                    // chain metadata for a block with Orchard actions.
```

Comment:
```suggestion
                    // that the subtraction will not underflow; if it would do so, we were given
                    // invalid chain metadata for a block with Orchard actions.
```